x/stake/keeper/delegation.go

```diff
@@ -0,0 +1,230 @@
+package keeper
+
+import (
+	"bytes"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/cosmos-sdk/x/stake/types"
+)
+
+// load a delegation
+func (k Keeper) GetDelegation(ctx sdk.Context,
+	delegatorAddr, validatorAddr sdk.Address) (delegation types.Delegation, found bool) {
+
+	store := ctx.KVStore(k.storeKey)
+	delegatorBytes := store.Get(GetDelegationKey(delegatorAddr, validatorAddr, k.cdc))
+	if delegatorBytes == nil {
+		return delegation, false
+	}
+
+	k.cdc.MustUnmarshalBinary(delegatorBytes, &delegation)
+	return delegation, true
+}
+
+// load all delegations used during genesis dump
+func (k Keeper) GetAllDelegations(ctx sdk.Context) (delegations []types.Delegation) {
+	store := ctx.KVStore(k.storeKey)
+	iterator := sdk.KVStorePrefixIterator(store, DelegationKey)
+
+	i := 0
+	for ; ; i++ {
+		if !iterator.Valid() {
+			iterator.Close()
+			break
+		}
+		bondBytes := iterator.Value()
+		var delegation types.Delegation
+		k.cdc.MustUnmarshalBinary(bondBytes, &delegation)
+		delegations = append(delegations, delegation)
+		iterator.Next()
+	}
+	return delegations[:i] // trim
+}
+
+// load all delegations for a delegator
+func (k Keeper) GetDelegations(ctx sdk.Context, delegator sdk.Address,
+	maxRetrieve int16) (delegations []types.Delegation) {
+
+	store := ctx.KVStore(k.storeKey)
+	delegatorPrefixKey := GetDelegationsKey(delegator, k.cdc)
+	iterator := sdk.KVStorePrefixIterator(store, delegatorPrefixKey) //smallest to largest
+
+	delegations = make([]types.Delegation, maxRetrieve)
+	i := 0
+	for ; ; i++ {
+		if !iterator.Valid() || i > int(maxRetrieve-1) {
+			iterator.Close()
```

Comment:
Can this be closed outside of the loop, per #1125

Reply:
addressed